nemo/collections/common/prompts/canary2.py

```diff
@@ -0,0 +1,206 @@
+        # It maps values provided in trcfg to the actual special tokens
+        # expected to be present in canary prompt.
+        # It used to be done in prompt_format_fn inside Dataset class corresponding to Canary,
+        # but we are not using it here anymore.
+        # This maps things such as '|task|: "asr"' to '|TASK|: "<|transcribe|>"'.
+        slot_values = map_manifest_values_to_special_tokens(slot_values)
+        return super().encode_turn(
+            prompt_template=prompt_template, expected_slots=expected_slots, slot_values=slot_values
+        )
+
+
+def map_manifest_values_to_special_tokens(slot_values: dict[str, str]) -> dict[str, str]:
+    slot_values = slot_values.copy()
+
+    any_special_token_present = False
+
+    lang_dict_compat = {"en": "en-US", "es": "es-ES", "fr": "fr-FR", "de": "de-DE"}
+    for k in ("source_lang", "target_lang"):
+        if k in slot_values and not ((v := slot_values[k]).startswith("<|") and v.endswith("|>")):
+            val = slot_values[k]
+            val = lang_dict_compat.get(val, val)
+            slot_values[k] = "<|" + val + "|>"
+            any_special_token_present = True
+
+    # Handle boolean slots
+    for k in ("pnc", "itn", "timestamp", "diarize"):
+        true_token = f"<|{k}|>"
+        false_token = f"<|no{k}|>"
+        if k in slot_values and slot_values[k] not in (true_token, false_token):
+            slot_values[k] = true_token if slot_values[k] in ("yes", "1", "True", "true", k) else false_token
+            any_special_token_present = True
+
+    # Auto-inject which tokenizer to look up in CanaryTokenizer if not provided,
+    # and slots for this turn correspond to user prompt.
+    if any_special_token_present and PromptFormatter.PROMPT_LANGUAGE_SLOT not in slot_values:
+        slot_values[PromptFormatter.PROMPT_LANGUAGE_SLOT] = CANARY_SPECIAL_TOKENIZER
+    else:
+        if (l := slot_values.get(PromptFormatter.PROMPT_LANGUAGE_SLOT)) is not None:
+            slot_values[PromptFormatter.PROMPT_LANGUAGE_SLOT] = lang_dict_compat.get(l, l)
+
+    return slot_values
+
+
+@registered_prompt_format_fn
+def canary2(
+    cuts: CutSet, tokenizer: TokenizerSpec
+) -> tuple[list[torch.Tensor], list[torch.Tensor], list[torch.Tensor]]:
+    """
+    Prepend and append control tokens to the token sequence as per Canary 2.0 format.
+
+    The prompt format syntax is defined in :class:`Canary2PromptFormatter`
+    """
+    formatter = Canary2PromptFormatter(tokenizer)
+
+    prompts_with_answers, prompts, answers = [], [], []
+    for cut in cuts:
+        if isinstance(cut, MixedCut):
+            cut = cut._first_non_padding_cut
+        if not isinstance(cut, MonoCut):
+            raise TypeError(
+                f"Expected input audio to have a single channel (required MonoCut/MixedCut, but we received: {cut=})"
+            )
+
+        # first, validate the utterance
+        expected_slots = {"source_lang", "target_lang", "pnc"}
```

Comment:
how about `expected_slots = formatter.expected_slots` 

should we make all the slots expected ie. source_lang, target_lang, pnc, itn, timestamp, diarize and we will set the default tags in data config yaml if something is missing in manifest file ?

Reply:
it will require rebuilding manifests (or at least data configs) for existing data, a bit of extra headache that I'm not sure is worth the time. WDYT?

Reply:
it shouldn't require us to rebuild manifests. it should be doable using the `tags` we have in train_ds yaml config? 

atleast may be the first part i.e, `expected_slots = formatter.expected_slots` ?

Reply:
discussed offline - keeping the current code to hide the extra complexity of new canary2 slots for users that are not interested in them

Reply:
after further conversation - made pnc optional too